Unable to write to streams buffer under bidi-stream

**Describe the bug**
While doing benchmark test zilla got error.
test was bidi-stream 15 connection and each connection send 100 messages.
for this test I'm using ghz benchmark cmd tool.
my ghz configuration is under description.
```json
{
    "insecure": true,
    "skipTLS": true,
    "proto": "./test.proto",
    "call": "test.TestService/SendTestData",
    "stream-interval": "5ms",
    "total": 15,
    "concurrency": 15,
    "data-file": "./messages.json",
    "metadata": {
        "trace_id": "{{.RequestNumber}}",
        "timestamp": "{{.TimestampUnix}}"
    },
    "host": "localhost:8082"
}
```

Total request is 15 and concurrency request is 15, each request client send 100 messages. thus, 1500 messages should be sended.
Zilla configuration file is almost same as grpc-kafka echo example. proto file is not that important (i guess). it's just like EchoBidiStream. 

```yaml
name: gRPC-TEST
bindings:

# Gateway ingress config
  tcp_server:
    type: tcp
    kind: server
    options:
      host: 0.0.0.0
      port: 8082
    exit: http_server
  http_server:
    type: http
    kind: server
    routes:
      - when:
          - headers:
              :scheme: http
              :authority: localhost:8082
        exit: grpc_server

# gRPC service definition
  grpc_server:
    type: grpc
    kind: server
    options:
      services:
        - proto/test.proto
    routes:
      - when:
          - method: test.TestService/*
        exit: grpc_kafka

# Proxy a gRPC service to a Kafka topic
  grpc_kafka:
    type: grpc-kafka
    kind: proxy
    options:
      correlation:
        headers:
          service: zilla:service
          method: zilla:method
          correlation-id: zilla:correlation-id
          reply-to: zilla:reply-to
    routes:
      - when:
          - method: test.TestService/*
        exit: kafka_cache_client
        with:
          capability: produce
          topic: echo-messages
          acks: leader_only
          reply-to: echo-messages

# Kafka caching layer
  kafka_cache_client:
    type: kafka
    kind: cache_client
    options:
      bootstrap:
        - echo-messages
    exit: kafka_cache_server
  kafka_cache_server:
    type: kafka
    kind: cache_server
    exit: kafka_client

# Connect to local Kafka
  kafka_client:
    type: kafka
    kind: client
    exit: kafka_tcp_client
  kafka_tcp_client:
    type: tcp
    kind: client
    options:
      host: kafka
      port: 9092
    routes:
      - when:
          - cidr: 0.0.0.0/0
```
**To Reproduce**
For reproducing this error you can follow under description.
1. make your own grpc-kafka echo server configuration or just use above configuration. (you can replace .proto file another echo proto file)
2. install ghz cmd tool.
3. use above the ghz config file (config.json)  but you should make messages.json for stream data. (or you can use python code that i wrote in bottom of this issue)
4. command `ghz --config config.json`
5. then you can see the error (Unable to write to streams buffer or org.agrona.concurrent.AgentTerminationException: java.lang.IndexOutOfBoundsException: index=65519 length=179 capacity=65536)

**Expected behavior**
Expected behavior is even if It's slow process the request correctly.

**Desktop (please complete the following information):**
 - OS: macbook pro 2019, intel
 - docker desktop, resources are cpu: 8, memory: 12GB, engine: v20.10.21

**Additional context**
Add any other context about the problem here.

1. Unable to write to streams buffer
```
2023-08-17 00:23:50 org.agrona.concurrent.AgentTerminationException: java.lang.IllegalStateException: Unable to write to streams buffer
2023-08-17 00:23:50     at io.aklivity.zilla.runtime.engine@0.9.51/io.aklivity.zilla.runtime.engine.internal.registry.DispatchAgent.doWork(DispatchAgent.java:681)
2023-08-17 00:23:50     at org.agrona.core/org.agrona.concurrent.AgentRunner.doDutyCycle(AgentRunner.java:291)
2023-08-17 00:23:50     at org.agrona.core/org.agrona.concurrent.AgentRunner.run(AgentRunner.java:164)
2023-08-17 00:23:50     at java.base/java.lang.Thread.run(Thread.java:833)
2023-08-17 00:23:50 Caused by: java.lang.IllegalStateException: Unable to write to streams buffer
2023-08-17 00:23:50     at io.aklivity.zilla.runtime.engine@0.9.51/io.aklivity.zilla.runtime.engine.internal.stream.Target.handleWrite(Target.java:159)
2023-08-17 00:23:50     at io.aklivity.zilla.runtime.binding.grpc.kafka@0.9.51/io.aklivity.zilla.runtime.binding.grpc.kafka.internal.stream.GrpcKafkaProxyFactory.doData(GrpcKafkaProxyFactory.java:2140)
2023-08-17 00:23:50     at io.aklivity.zilla.runtime.binding.grpc.kafka@0.9.51/io.aklivity.zilla.runtime.binding.grpc.kafka.internal.stream.GrpcKafkaProxyFactory$KafkaProduceProxy.doKafkaData(GrpcKafkaProxyFactory.java:1538)
2023-08-17 00:23:50     at io.aklivity.zilla.runtime.binding.grpc.kafka@0.9.51/io.aklivity.zilla.runtime.binding.grpc.kafka.internal.stream.GrpcKafkaProxyFactory$GrpcProduceProxy.onGrpcData(GrpcKafkaProxyFactory.java:1172)
2023-08-17 00:23:50     at io.aklivity.zilla.runtime.binding.grpc.kafka@0.9.51/io.aklivity.zilla.runtime.binding.grpc.kafka.internal.stream.GrpcKafkaProxyFactory$GrpcProduceProxy.onGrpcMessage(GrpcKafkaProxyFactory.java:1086)
2023-08-17 00:23:50     at io.aklivity.zilla.runtime.binding.grpc.kafka@0.9.51/io.aklivity.zilla.runtime.binding.grpc.kafka.internal.stream.GrpcKafkaProxyFactory.lambda$newStream$1(GrpcKafkaProxyFactory.java:228)
2023-08-17 00:23:50     at io.aklivity.zilla.runtime.engine@0.9.51/io.aklivity.zilla.runtime.engine.internal.registry.DispatchAgent.handleReadInitial(DispatchAgent.java:1038)
2023-08-17 00:23:50     at io.aklivity.zilla.runtime.engine@0.9.51/io.aklivity.zilla.runtime.engine.internal.registry.DispatchAgent.handleRead(DispatchAgent.java:1005)
2023-08-17 00:23:50     at io.aklivity.zilla.runtime.engine@0.9.51/io.aklivity.zilla.runtime.engine.internal.concurent.ManyToOneRingBuffer.read(ManyToOneRingBuffer.java:181)
2023-08-17 00:23:50     at io.aklivity.zilla.runtime.engine@0.9.51/io.aklivity.zilla.runtime.engine.internal.registry.DispatchAgent.doWork(DispatchAgent.java:675)
2023-08-17 00:23:50     ... 3 more
2023-08-17 00:23:50     Suppressed: java.lang.Exception: [engine/data#4]        [0x0404000000000029] streams=[consumeAt=0x0004c350 (0x000000000014c350), produceAt=0x000fff18 (0x00000000001fff18)]
2023-08-17 00:23:50             at io.aklivity.zilla.runtime.engine@0.9.51/io.aklivity.zilla.runtime.engine.internal.registry.DispatchAgent.doWork(DispatchAgent.java:679)
2023-08-17 00:23:50             ... 3 more
```
2. IndexOutOfBoundsException
```
2023-08-17 00:23:33 org.agrona.concurrent.AgentTerminationException: java.lang.IndexOutOfBoundsException: index=65519 length=179 capacity=65536
2023-08-17 00:23:33     at io.aklivity.zilla.runtime.engine@0.9.51/io.aklivity.zilla.runtime.engine.internal.registry.DispatchAgent.doWork(DispatchAgent.java:681)
2023-08-17 00:23:33     at org.agrona.core/org.agrona.concurrent.AgentRunner.doDutyCycle(AgentRunner.java:291)
2023-08-17 00:23:33     at org.agrona.core/org.agrona.concurrent.AgentRunner.run(AgentRunner.java:164)
2023-08-17 00:23:33     at java.base/java.lang.Thread.run(Thread.java:833)
2023-08-17 00:23:33 Caused by: java.lang.IndexOutOfBoundsException: index=65519 length=179 capacity=65536
2023-08-17 00:23:33     at org.agrona.core/org.agrona.concurrent.UnsafeBuffer.boundsCheck0(UnsafeBuffer.java:1692)
2023-08-17 00:23:33     at org.agrona.core/org.agrona.concurrent.UnsafeBuffer.putBytes(UnsafeBuffer.java:945)
2023-08-17 00:23:33     at io.aklivity.zilla.runtime.binding.kafka@0.9.51/io.aklivity.zilla.runtime.binding.kafka.internal.stream.KafkaClientProduceFactory$KafkaProduceStream$KafkaProduceClient.doEncodeRecordFin(KafkaClientProduceFactory.java:1735)
2023-08-17 00:23:33     at io.aklivity.zilla.runtime.binding.kafka@0.9.51/io.aklivity.zilla.runtime.binding.kafka.internal.stream.KafkaClientProduceFactory.flushRecordContFin(KafkaClientProduceFactory.java:576)
2023-08-17 00:23:33     at io.aklivity.zilla.runtime.binding.kafka@0.9.51/io.aklivity.zilla.runtime.binding.kafka.internal.stream.KafkaClientProduceFactory$KafkaProduceStream$KafkaProduceClient.flush(KafkaClientProduceFactory.java:1581)
2023-08-17 00:23:33     at io.aklivity.zilla.runtime.binding.kafka@0.9.51/io.aklivity.zilla.runtime.binding.kafka.internal.stream.KafkaClientProduceFactory$KafkaProduceStream.onApplicationData(KafkaClientProduceFactory.java:979)
2023-08-17 00:23:33     at io.aklivity.zilla.runtime.binding.kafka@0.9.51/io.aklivity.zilla.runtime.binding.kafka.internal.stream.KafkaClientProduceFactory$KafkaProduceStream.onApplication(KafkaClientProduceFactory.java:903)
2023-08-17 00:23:33     at io.aklivity.zilla.runtime.binding.kafka@0.9.51/io.aklivity.zilla.runtime.binding.kafka.internal.stream.KafkaClientProduceFactory.lambda$newStream$1(KafkaClientProduceFactory.java:257)
2023-08-17 00:23:33     at io.aklivity.zilla.runtime.engine@0.9.51/io.aklivity.zilla.runtime.engine.internal.registry.DispatchAgent.handleReadInitial(DispatchAgent.java:1038)
2023-08-17 00:23:33     at io.aklivity.zilla.runtime.engine@0.9.51/io.aklivity.zilla.runtime.engine.internal.registry.DispatchAgent.handleRead(DispatchAgent.java:1005)
2023-08-17 00:23:33     at io.aklivity.zilla.runtime.engine@0.9.51/io.aklivity.zilla.runtime.engine.internal.concurent.ManyToOneRingBuffer.read(ManyToOneRingBuffer.java:181)
2023-08-17 00:23:33     at io.aklivity.zilla.runtime.engine@0.9.51/io.aklivity.zilla.runtime.engine.internal.registry.DispatchAgent.doWork(DispatchAgent.java:675)
2023-08-17 00:23:33     ... 3 more
2023-08-17 00:23:33     Suppressed: java.lang.Exception: [engine/data#7]        [0x070700000000000b] streams=[consumeAt=0x0001f5f0 (0x000000000021f5f0), produceAt=0x00081080 (0x0000000000281080)]
2023-08-17 00:23:33             at io.aklivity.zilla.runtime.engine@0.9.51/io.aklivity.zilla.runtime.engine.internal.registry.DispatchAgent.doWork(DispatchAgent.java:679)
2023-08-17 00:23:33             ... 3 more
```
3. proto file
```
syntax = "proto3";
package test;

message TestRequest{
  message TestOneRequest {
    int64 id = 1;
  }

  message TestTwoRequest {
    string message = 1;
  }

  oneof message {
    TestOneRequest one = 1;
    TestTwoRequest two = 2;
  }
}

service TestService {
  rpc SendTestData(stream TestRequest) returns (stream TestRequest);
}

```

4. this python code is for create stream data.
```python
import json

message_list = []

for i in range(1, 101):
    message = {
        "one": {
            "id": i
        }
    }
    message_list.append(message)


json_messages = json.dumps(message_list, indent=4)

with open('messages.json', 'w') as file:
    file.write(json_messages)

```

--- a/runtime/binding-kafka/src/main/java/io/aklivity/zilla/runtime/binding/kafka/internal/KafkaConfiguration.java
+++ b/runtime/binding-kafka/src/main/java/io/aklivity/zilla/runtime/binding/kafka/internal/KafkaConfiguration.java
@@ -47,6 +47,7 @@ public class KafkaConfiguration extends Configuration
     public static final IntPropertyDef KAFKA_CLIENT_PRODUCE_MAX_REQUEST_MILLIS;
     public static final IntPropertyDef KAFKA_CLIENT_PRODUCE_MAX_RESPONSE_MILLIS;
     public static final IntPropertyDef KAFKA_CLIENT_PRODUCE_MAX_BYTES;
+    public static final IntPropertyDef KAFKA_CLIENT_PRODUCE_RECORD_FRAMING_SIZE;
     public static final PropertyDef<Path> KAFKA_CACHE_DIRECTORY;
     public static final LongPropertyDef KAFKA_CACHE_PRODUCE_CAPACITY;
     public static final PropertyDef<KafkaCacheCleanupPolicy> KAFKA_CACHE_CLEANUP_POLICY;
@@ -88,6 +89,7 @@ public class KafkaConfiguration extends Configuration
         KAFKA_CLIENT_PRODUCE_MAX_REQUEST_MILLIS = config.property("client.produce.max.request.millis", 0);
         KAFKA_CLIENT_PRODUCE_MAX_RESPONSE_MILLIS = config.property("client.produce.max.response.millis", 120000);
         KAFKA_CLIENT_PRODUCE_MAX_BYTES = config.property("client.produce.max.bytes", Integer.MAX_VALUE);
+        KAFKA_CLIENT_PRODUCE_RECORD_FRAMING_SIZE = config.property("client.produce.record.framing.size", 512);
         KAFKA_CLIENT_SASL_SCRAM_NONCE = config.property(NonceSupplier.class, "client.sasl.scram.nonce",
             KafkaConfiguration::decodeNonceSupplier, KafkaConfiguration::defaultNonceSupplier);
         KAFKA_CLIENT_GROUP_REBALANCE_TIMEOUT = config.property(Duration.class, "client.group.rebalance.timeout",
@@ -170,6 +172,11 @@ public class KafkaConfiguration extends Configuration
     public int clientProduceMaxBytes()
     {
         return KAFKA_CLIENT_PRODUCE_MAX_BYTES.getAsInt(this);
+    }
+
+    public int clientProduceRecordFramingSize()
+    {
+        return KAFKA_CLIENT_PRODUCE_RECORD_FRAMING_SIZE.getAsInt(this);
     }
 
     public Path cacheDirectory()

--- a/runtime/binding-kafka/src/main/java/io/aklivity/zilla/runtime/binding/kafka/internal/stream/KafkaClientProduceFactory.java
+++ b/runtime/binding-kafka/src/main/java/io/aklivity/zilla/runtime/binding/kafka/internal/stream/KafkaClientProduceFactory.java
@@ -86,7 +86,7 @@ public final class KafkaClientProduceFactory extends KafkaClientSaslHandshaker i
 {
     private static final int PRODUCE_REQUEST_RECORDS_OFFSET_MAX = 512;
 
-    private static final int KAFKA_RECORD_FRAMING = 100; // TODO
+    private static final int KAFKA_RECORD_FRAMING = 512; // TODO
 
     private static final int FLAGS_CON = 0x00;
     private static final int FLAGS_FIN = 0x01;
@@ -538,6 +538,7 @@ public final class KafkaClientProduceFactory extends KafkaClientSaslHandshaker i
         client.valueChecksum = kafkaProduceDataEx.crc32c();
         final int valueSize = payload != null ? payload.sizeof() : 0;
         client.valueCompleteSize = valueSize + client.encodeableRecordBytesDeferred;
+
 
         final int maxEncodeableBytes = client.encodeSlotLimit + client.valueCompleteSize + KAFKA_RECORD_FRAMING;
         if (client.encodeSlot != NO_SLOT &&
@@ -1191,6 +1192,7 @@ public final class KafkaClientProduceFactory extends KafkaClientSaslHandshaker i
             private int encodeableRecordCount;
             private int encodeableRecordBytes;
             private int encodeableRecordBytesDeferred;
+            private int encodeableRecordValueBytes;
             private int flushableRequestBytes;
 
             private int decodeSlot = NO_SLOT;
@@ -1652,6 +1654,7 @@ public final class KafkaClientProduceFactory extends KafkaClientSaslHandshaker i
 
                 encodeSlotBuffer.putBytes(encodeSlotLimit, encodeBuffer, 0, encodeProgress);
                 encodeSlotLimit += encodeProgress;
+                encodeableRecordValueBytes = 0;
 
                 if (headersCount > 0)
                 {
@@ -1689,6 +1692,7 @@ public final class KafkaClientProduceFactory extends KafkaClientSaslHandshaker i
 
                     encodeSlotBuffer.putBytes(encodeSlotLimit,  value.buffer(), value.offset(), length);
                     encodeSlotLimit += length;
+                    encodeableRecordValueBytes += length;
 
                     if ((flags & FLAGS_FIN) == 0)
                     {
@@ -1893,7 +1897,8 @@ public final class KafkaClientProduceFactory extends KafkaClientSaslHandshaker i
 
                 final ByteBuffer encodeSlotByteBuffer = encodePool.byteBuffer(encodeSlot);
                 final int encodeSlotBytePosition = encodeSlotByteBuffer.position();
-                encodeSlotByteBuffer.limit(encodeSlotBytePosition + encodeSlotLimit);
+                final int partialValueSize = flushFlags != FLAGS_FIN ? encodeableRecordValueBytes : 0;
+                encodeSlotByteBuffer.limit(encodeSlotBytePosition + encodeSlotLimit - partialValueSize);
                 encodeSlotByteBuffer.position(encodeSlotBytePosition + encodeSlotOffset + crcLimit);
 
                 final CRC32C crc = crc32c;

--- a/runtime/binding-kafka/src/main/java/io/aklivity/zilla/runtime/binding/kafka/internal/stream/KafkaClientProduceFactory.java
+++ b/runtime/binding-kafka/src/main/java/io/aklivity/zilla/runtime/binding/kafka/internal/stream/KafkaClientProduceFactory.java
@@ -86,8 +86,6 @@ public final class KafkaClientProduceFactory extends KafkaClientSaslHandshaker i
 {
     private static final int PRODUCE_REQUEST_RECORDS_OFFSET_MAX = 512;
 
-    private static final int KAFKA_RECORD_FRAMING = 512; // TODO
-
     private static final int FLAGS_CON = 0x00;
     private static final int FLAGS_FIN = 0x01;
     private static final int FLAGS_INIT = 0x02;
@@ -178,6 +176,7 @@ public final class KafkaClientProduceFactory extends KafkaClientSaslHandshaker i
     private final KafkaProduceClientDecoder decodeReject = this::decodeReject;
 
     private final int produceMaxWaitMillis;
+    private final int produceRecordFramingSize;
     private final long produceRequestMaxDelay;
     private final int kafkaTypeId;
     private final int proxyTypeId;
@@ -201,6 +200,7 @@ public final class KafkaClientProduceFactory extends KafkaClientSaslHandshaker i
     {
         super(config, context);
         this.produceMaxWaitMillis = config.clientProduceMaxResponseMillis();
+        this.produceRecordFramingSize = config.clientProduceRecordFramingSize();
         this.produceRequestMaxDelay = config.clientProduceMaxRequestMillis();
         this.kafkaTypeId = context.supplyTypeId(KafkaBinding.NAME);
         this.proxyTypeId = context.supplyTypeId("proxy");
@@ -540,7 +540,7 @@ public final class KafkaClientProduceFactory extends KafkaClientSaslHandshaker i
         client.valueCompleteSize = valueSize + client.encodeableRecordBytesDeferred;
 
 
-        final int maxEncodeableBytes = client.encodeSlotLimit + client.valueCompleteSize + KAFKA_RECORD_FRAMING;
+        final int maxEncodeableBytes = client.encodeSlotLimit + client.valueCompleteSize + produceRecordFramingSize;
         if (client.encodeSlot != NO_SLOT &&
             maxEncodeableBytes > encodePool.slotCapacity())
         {
@@ -970,7 +970,8 @@ public final class KafkaClientProduceFactory extends KafkaClientSaslHandshaker i
 
             assert initialAck <= initialSeq;
 
-            if (initialSeq > initialAck + initialMax)
+            if (initialSeq > initialAck + initialMax ||
+                extension.sizeof() > produceRecordFramingSize)
             {
                 cleanupApplication(traceId, EMPTY_OCTETS);
                 client.cleanupNetwork(traceId);
@@ -1095,7 +1096,7 @@ public final class KafkaClientProduceFactory extends KafkaClientSaslHandshaker i
                 state = KafkaState.openedInitial(state);
 
                 doWindow(application, originId, routedId, initialId, initialSeq, initialAck, initialMax,
-                        traceId, client.authorization, client.initialBud, KAFKA_RECORD_FRAMING);
+                        traceId, client.authorization, client.initialBud, produceRecordFramingSize);
             }
         }
 
@@ -1569,7 +1570,8 @@ public final class KafkaClientProduceFactory extends KafkaClientSaslHandshaker i
                 final int length = limit - progress;
                 if (encodeSlot != NO_SLOT &&
                     flushableRequestBytes > 0 &&
-                    encodeSlotLimit + length + KAFKA_RECORD_FRAMING + flushableRecordHeadersBytes > encodePool.slotCapacity())
+                    encodeSlotLimit + length + produceRecordFramingSize + flushableRecordHeadersBytes >
+                        encodePool.slotCapacity())
                 {
                     doNetworkData(traceId, budgetId, EMPTY_BUFFER, 0, 0);
                 }
@@ -1681,7 +1683,8 @@ public final class KafkaClientProduceFactory extends KafkaClientSaslHandshaker i
                 {
                     final int length = value.sizeof();
 
-                    final int encodeableBytes = KAFKA_RECORD_FRAMING + encodeSlotLimit + length + flushableRecordHeadersBytes;
+                    final int encodeableBytes = produceRecordFramingSize + encodeSlotLimit +
+                        length + flushableRecordHeadersBytes;
                     if (encodeableBytes >= encodePool.slotCapacity())
                     {
                         doEncodeRequestIfNecessary(traceId, budgetId);

--- a/runtime/binding-kafka/src/test/java/io/aklivity/zilla/runtime/binding/kafka/internal/stream/ClientMergedIT.java
+++ b/runtime/binding-kafka/src/test/java/io/aklivity/zilla/runtime/binding/kafka/internal/stream/ClientMergedIT.java
@@ -49,7 +49,7 @@ public class ClientMergedIT
         .countersBufferCapacity(8192)
         .configure(ENGINE_BUFFER_SLOT_CAPACITY, 8192)
         .configure(KAFKA_CLIENT_META_MAX_AGE_MILLIS, 1000)
-        .configure(KAFKA_CLIENT_PRODUCE_MAX_BYTES, 116)
+        .configure(KAFKA_CLIENT_PRODUCE_MAX_BYTES, 528)
         .configurationRoot("io/aklivity/zilla/specs/binding/kafka/config")
         .external("net0")
         .clean();
@@ -234,7 +234,7 @@ public class ClientMergedIT
     @Configure(
         name = "zilla.binding.kafka.client.produce.max.bytes",
         value = "200000")
-    @ScriptProperty("padding ${512 + 100}")
+    @ScriptProperty("padding ${512 + 512}")
     public void shouldProduceMergedMessageValue10k() throws Exception
     {
         k3po.finish();
@@ -248,7 +248,7 @@ public class ClientMergedIT
     @Configure(
         name = "zilla.binding.kafka.client.produce.max.bytes",
         value = "200000")
-    @ScriptProperty("padding ${512 + 100}")
+    @ScriptProperty("padding ${512 + 512}")
     public void shouldProduceMergedMessageValue100k() throws Exception
     {
         k3po.finish();

--- a/specs/binding-kafka.spec/src/main/scripts/io/aklivity/zilla/specs/binding/kafka/streams/application/produce/message.value.100k/client.rpt
+++ b/specs/binding-kafka.spec/src/main/scripts/io/aklivity/zilla/specs/binding/kafka/streams/application/produce/message.value.100k/client.rpt
@@ -73,7 +73,7 @@ read zilla:begin.ext ${kafka:beginEx()
 write zilla:data.ext ${kafka:dataEx()
                               .typeId(zilla:id("kafka"))
                               .produce()
-                                  .deferred(102400 - 8192 + 512 + 100)
+                                  .deferred(102400 - 8192 + 512 + 512)
                                   .timestamp(newTimestamp)
                                   .build()
                               .build()}

--- a/specs/binding-kafka.spec/src/main/scripts/io/aklivity/zilla/specs/binding/kafka/streams/application/produce/message.value.100k/server.rpt
+++ b/specs/binding-kafka.spec/src/main/scripts/io/aklivity/zilla/specs/binding/kafka/streams/application/produce/message.value.100k/server.rpt
@@ -18,7 +18,7 @@ property serverAddress "zilla://streams/app0"
 
 accept ${serverAddress}
     option zilla:window 8192
-    option zilla:padding 612
+    option zilla:padding 1024
     option zilla:transmission "half-duplex"
 
 accepted
@@ -71,7 +71,7 @@ write zilla:begin.ext ${kafka:beginEx()
 read zilla:data.ext ${kafka:matchDataEx()
                              .typeId(zilla:id("kafka"))
                              .produce()
-                                 .deferred(102400 - 8192 + 512 + 100)
+                                 .deferred(102400 - 8192 + 512 + 512)
                                  .build()
                              .build()}
 

--- a/specs/binding-kafka.spec/src/main/scripts/io/aklivity/zilla/specs/binding/kafka/streams/application/produce/message.value.10k/client.rpt
+++ b/specs/binding-kafka.spec/src/main/scripts/io/aklivity/zilla/specs/binding/kafka/streams/application/produce/message.value.10k/client.rpt
@@ -73,7 +73,7 @@ read zilla:begin.ext ${kafka:beginEx()
 write zilla:data.ext ${kafka:dataEx()
                               .typeId(zilla:id("kafka"))
                               .produce()
-                                  .deferred(10240 - 8192 + 512 + 100)
+                                  .deferred(10240 - 8192 + 512 + 512)
                                   .timestamp(newTimestamp)
                                   .build()
                               .build()}

--- a/specs/binding-kafka.spec/src/main/scripts/io/aklivity/zilla/specs/binding/kafka/streams/application/produce/message.value.10k/server.rpt
+++ b/specs/binding-kafka.spec/src/main/scripts/io/aklivity/zilla/specs/binding/kafka/streams/application/produce/message.value.10k/server.rpt
@@ -18,7 +18,7 @@ property serverAddress "zilla://streams/app0"
 
 accept ${serverAddress}
     option zilla:window 8192
-    option zilla:padding 612
+    option zilla:padding 1024
     option zilla:transmission "half-duplex"
 
 accepted
@@ -71,7 +71,7 @@ write zilla:begin.ext ${kafka:beginEx()
 read zilla:data.ext ${kafka:matchDataEx()
                              .typeId(zilla:id("kafka"))
                              .produce()
-                                 .deferred(10240 - 8192 + 512 + 100)
+                                 .deferred(10240 - 8192 + 512 + 512)
                                  .build()
                              .build()}
 read zilla:data.ext ${kafka:matchDataEx()

--- a/specs/binding-kafka.spec/src/main/scripts/io/aklivity/zilla/specs/binding/kafka/streams/application/produce/message.values.sequential/client.rpt
+++ b/specs/binding-kafka.spec/src/main/scripts/io/aklivity/zilla/specs/binding/kafka/streams/application/produce/message.values.sequential/client.rpt
@@ -81,7 +81,7 @@ write zilla:data.ext ${kafka:dataEx()
                               .produce()
                                   .build()
                               .build()}
-write ${kafka:randomBytes(7580)}
+write ${kafka:randomBytes(8192-(512+512))}
 write flush
 
 write zilla:data.ext ${kafka:dataEx()

--- a/specs/binding-kafka.spec/src/main/scripts/io/aklivity/zilla/specs/binding/kafka/streams/application/produce/message.values.sequential/server.rpt
+++ b/specs/binding-kafka.spec/src/main/scripts/io/aklivity/zilla/specs/binding/kafka/streams/application/produce/message.values.sequential/server.rpt
@@ -79,7 +79,7 @@ read zilla:data.ext ${kafka:matchDataEx()
                              .produce()
                                  .build()
                              .build()}
-read [0..7580]
+read [0..7168]
 
 read zilla:data.ext ${kafka:matchDataEx()
                              .typeId(zilla:id("kafka"))

--- a/specs/binding-kafka.spec/src/main/scripts/io/aklivity/zilla/specs/binding/kafka/streams/network/produce.v3/message.values.sequential/client.rpt
+++ b/specs/binding-kafka.spec/src/main/scripts/io/aklivity/zilla/specs/binding/kafka/streams/network/produce.v3/message.values.sequential/client.rpt
@@ -78,7 +78,7 @@ write zilla:begin.ext ${proxy:beginEx()
 
 connected
 
-write 7690
+write 7278
       0s
       3s
       ${newRequestId}
@@ -90,9 +90,9 @@ write 7690
       4s "test"
       1
       0
-      7650                               # record set size
+      7238                               # record set size
       0L                                 # first offset
-      7638                               # length
+      7226                               # length
       -1
       [0x02]
       0x4e8723aa
@@ -104,13 +104,13 @@ write 7690
       -1s
       -1
       1                                  # records
-      ${kafka:varint(7587)}
+      ${kafka:varint(7175)}
       [0x00]
       ${kafka:varint(0)}
       ${kafka:varint(0)}
       ${kafka:varint(-1)}                # key
-      ${kafka:varint(7580)}              # value
-      ${kafka:randomBytes(7580)}
+      ${kafka:varint(7168)}              # value
+      ${kafka:randomBytes(7168)}
       ${kafka:varint(0)}                 # headers
 
 read 44

--- a/specs/binding-kafka.spec/src/main/scripts/io/aklivity/zilla/specs/binding/kafka/streams/network/produce.v3/message.values.sequential/server.rpt
+++ b/specs/binding-kafka.spec/src/main/scripts/io/aklivity/zilla/specs/binding/kafka/streams/network/produce.v3/message.values.sequential/server.rpt
@@ -74,7 +74,7 @@ read zilla:begin.ext ${proxy:beginEx()
 
 connected
 
-read 7690
+read 7278
      0s
      3s
      (int:requestId)
@@ -86,9 +86,9 @@ read 7690
      4s "test"
      1
      0
-     7650                               # record set size
+     7238                               # record set size
      0L                                 # first offset
-     7638                               # length
+     7226                               # length
      -1
      [0x02]
      [0..4]
@@ -100,13 +100,13 @@ read 7690
      -1s
      -1
      1                                  # records
-     ${kafka:varint(7587)}
+     ${kafka:varint(7175)}
      [0x00]
      ${kafka:varint(0)}
      ${kafka:varint(0)}
      ${kafka:varint(-1)}                # key
-     ${kafka:varint(7580)}              # value
-     [0..7580]
+     ${kafka:varint(7168)}              # value
+     [0..7168]
      ${kafka:varint(0)}                 # headers
 
 write 44

